archinstall/lib/models/users.py

```diff
@@ -103,72 +107,150 @@ def _check_password_strength(
 		return PasswordStrength.VERY_WEAK
 
 
-_UserSerialization = TypedDict('_UserSerialization', {'username': str, '!password': str, 'sudo': bool})
+_UserSerialization = TypedDict(
+	'_UserSerialization',
+	{
+		'username': str,
+		'!password': NotRequired[str],
+		'sudo': bool,
+		'groups': list[str],
+		'enc_password': str | None
+	}
+)
 
 
-@dataclass
-class User:
-	username: str
-	password: str
-	sudo: bool
+class Password:
+	def __init__(
+		self,
+		plaintext: str = '',
+		enc_password: str | None = None
+	):
+		if plaintext:
+			enc_password = self._gen_yescrypt_hash(plaintext)
+
+		if not plaintext and not enc_password:
+			raise ValueError('Either plaintext or enc_password must be provided')
+
+		self._plaintext = plaintext
+		self.enc_password = enc_password
 
 	@property
-	def groups(self) -> list[str]:
-		# this property should be transferred into a class attr instead
-		# if it's every going to be used
-		return []
+	def plaintext(self) -> str:
+		return self._plaintext
 
-	def json(self) -> _UserSerialization:
-		return {
-			'username': self.username,
-			'!password': self.password,
-			'sudo': self.sudo
-		}
+	@plaintext.setter
+	def plaintext(self, value: str):
+		self._plaintext = value
+		self.enc_password = self._gen_yescrypt_hash(value)
 
-	@classmethod
-	def _parse(cls, config_users: list[_UserSerialization]) -> list['User']:
-		users = []
+	@override
+	def __eq__(self, other: object) -> bool:
+		if not isinstance(other, Password):
+			return NotImplemented
 
-		for entry in config_users:
-			username = entry.get('username', None)
-			password = entry.get('!password', '')
-			sudo = entry.get('sudo', False)
+		if self._plaintext and other._plaintext:
+			return self._plaintext == other._plaintext
 
-			if username is None:
-				continue
+		return self.enc_password == other.enc_password
 
-			user = User(username, password, sudo)
-			users.append(user)
+	def hidden(self) -> str:
+		if self._plaintext:
+			return '*' * len(self._plaintext)
+		else:
+			return '*' * 8
 
-		return users
+	@cached_property
+	def _libcrypt(self):
+		libc = ctypes.CDLL("libcrypt.so")
 
-	@classmethod
-	def _parse_backwards_compatible(cls, config_users: dict[str, dict[str, str]], sudo: bool) -> list['User']:
-		if len(config_users.keys()) > 0:
-			username = list(config_users.keys())[0]
-			password = config_users[username]['!password']
+		libc.crypt.argtypes = [ctypes.c_char_p, ctypes.c_char_p]
+		libc.crypt.restype = ctypes.c_char_p
+
+		libc.crypt_gensalt.argtypes = [ctypes.c_char_p, ctypes.c_int, ctypes.c_void_p, ctypes.c_size_t]
+		libc.crypt_gensalt.restype = ctypes.c_char_p
+
+		return libc
+
+	def _gen_yescrypt_salt(self) -> bytes:
+		salt_prefix = b'$y$'
+		rounds = 9
+		entropy = os.urandom(64)
+
+		salt = self._libcrypt.crypt_gensalt(salt_prefix, rounds, entropy, 64)
```

Comment:
From [`crypt_gensalt(3)`](https://man.archlinux.org/man/crypt_gensalt.3):

> char *
**crypt_gensalt**(_const char *prefix, unsigned long count, const char *rbytes, int nrbytes_);

> If rbytes is a null pointer, an appropriate number of random bytes will be obtained from the operating system, and nrbytes is ignored.

Reply:
Updated the call to generate the salt